File: api/src/models/FuelLog.ts
import mongoose, { Schema, Types, Document } from 'mongoose';

// Interface for FuelLog document
export interface IFuelLog extends Document {
  vehicleId: Types.ObjectId;
  date: Date;
  mileage: number;
  fuelAmount: number;
  totalCost: number;
  notes?: string;
  createdAt: Date;
  updatedAt: Date;
}

// Interface for static methods
interface IFuelLogModel extends mongoose.Model<IFuelLog> {
  findByVehicle(vehicleId: string, options?: { page?: number; limit?: number }): Promise<IFuelLog[]>;
  findByVehicleAndId(vehicleId: string, fuelLogId: string): Promise<IFuelLog | null>;
  countByVehicle(vehicleId: string): Promise<number>;
}

// Schema definition
const FuelLogSchema = new Schema<IFuelLog, IFuelLogModel>(
  {
    vehicleId: {
      type: Schema.Types.ObjectId,
      ref: 'Vehicle',
      required: [true, 'Vehicle ID is required'],
      index: true,
    },
    date: {
      type: Date,
      required: [true, 'Date is required'],
      default: Date.now,
      index: true,
    },
    mileage: {
      type: Number,
      required: [true, 'Mileage is required'],
      min: [0, 'Mileage cannot be negative'],
    },
    fuelAmount: {
      type: Number,
      required: [true, 'Fuel amount is required'],
      min: [0, 'Fuel amount cannot be negative'],
    },
    totalCost: {
      type: Number,
      required: [true, 'Total cost is required'],
      min: [0, 'Total cost cannot be negative'],
    },
    notes: {
      type: String,
      trim: true,
      maxlength: [500, 'Notes cannot exceed 500 characters'],
      default: undefined,
    },
  },
  {
    timestamps: true,
  }
);

// Compound index for vehicleId + date (for efficient querying and sorting)
FuelLogSchema.index({ vehicleId: 1, date: -1 });

// Statics (model-level)
FuelLogSchema.statics.findByVehicle = function (vehicleId: string, options?: { page?: number; limit?: number }) {
  let query = this.find({ vehicleId }).sort({ date: -1 });

  if (options && options.page !== undefined && options.limit !== undefined) {
    const skip = (options.page - 1) * options.limit;
    query = query.skip(skip).limit(options.limit);
  }

  return query;
};

FuelLogSchema.statics.findByVehicleAndId = function (vehicleId: string, fuelLogId: string) {
  return this.findOne({ _id: fuelLogId, vehicleId });
};

FuelLogSchema.statics.countByVehicle = function (vehicleId: string) {
  return this.countDocuments({ vehicleId });
};

// Virtuals
FuelLogSchema.virtual('id').get(function () {
  return this._id.toHexString();
});

// Transform JSON output
FuelLogSchema.set('toJSON', {
  virtuals: true,
  transform: function (_doc, ret) {
    const transformed = { ...ret } as Record<string, unknown>;
    delete transformed._id;
    delete transformed.__v;
    return transformed;
  },
});

export default mongoose.model<IFuelLog, IFuelLogModel>('FuelLog', FuelLogSchema);

File: mobile/lib/domain/entities/fuel_log.dart
class FuelLog {
  final String id;
  final String vehicleId;
  final DateTime date;
  final int mileage;
  final double fuelAmount;
  final double totalCost;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  const FuelLog({
    required this.id,
    required this.vehicleId,
    required this.date,
    required this.mileage,
    required this.fuelAmount,
    required this.totalCost,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  factory FuelLog.fromJson(Map<String, dynamic> json) => FuelLog(
        id: json['id'] as String,
        vehicleId: json['vehicleId'] as String,
        date: DateTime.parse(json['date'] as String),
        mileage: json['mileage'] as int,
        fuelAmount: (json['fuelAmount'] as num).toDouble(),
        totalCost: (json['totalCost'] as num).toDouble(),
        notes: json['notes'] as String?,
        createdAt: DateTime.parse(json['createdAt'] as String),
        updatedAt: DateTime.parse(json['updatedAt'] as String),
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'vehicleId': vehicleId,
        'date': date.toIso8601String(),
        'mileage': mileage,
        'fuelAmount': fuelAmount,
        'totalCost': totalCost,
        'notes': notes,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
      };

  FuelLog copyWith({
    String? id,
    String? vehicleId,
    DateTime? date,
    int? mileage,
    double? fuelAmount,
    double? totalCost,
    String? notes,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) =>
      FuelLog(
        id: id ?? this.id,
        vehicleId: vehicleId ?? this.vehicleId,
        date: date ?? this.date,
        mileage: mileage ?? this.mileage,
        fuelAmount: fuelAmount ?? this.fuelAmount,
        totalCost: totalCost ?? this.totalCost,
        notes: notes ?? this.notes,
        createdAt: createdAt ?? this.createdAt,
        updatedAt: updatedAt ?? this.updatedAt,
      );
}